### PR TITLE
Fix and improve parsing error output.

### PR DESCRIPTION
**Pull request by fabxc**

### CODE REVIEW

File: promql/lex.go [hunk 1]
```diff
@@ -422,7 +442,7 @@ func lexStatements(l *lexer) stateFn {
 			}
 		}
 		fallthrough
-	case isAlphaNumeric(r):
+	case unicode.IsLetter(r) || r == '_' || r == ':':
```
Comment (juliusv) [inline promql/lex.go]: This would allow all kinds of non-ASCII letters :P

File: promql/lex.go [hunk 2]
```diff
@@ -472,7 +492,7 @@ func lexInsideBraces(l *lexer) stateFn {
 		return l.errorf("unexpected EOF inside braces")
 	case isSpace(r):
 		return lexSpace
-	case isAlphaNumeric(r):
+	case unicode.IsLetter(r) || r == '_':
```
Comment (juliusv) [inline promql/lex.go]: dito

File: promql/parse.go [hunk 3]
```diff
@@ -101,7 +101,7 @@ func (p *parser) parseExpr() (expr Expr, err error) {
 			continue
 		}
 		if expr != nil {
-			p.errorf("expression read but input remaining")
+			p.errorf("remaining input %q could not be parsed", p.lex.input[p.lex.lastPos:])
```
Comment (juliusv) [inline promql/parse.go]: I'd put the remaining input at the end, in case it's long: "failed to parse additional input after expression: %q"

Comment (fabxc) [inline promql/parse.go]: How about `%.10q...`?

File: promql/parse.go [hunk 4]
```diff
@@ -174,29 +177,39 @@ func (p *parser) error(err error) {
 	panic(perr)
 }
 
+func tokenDesc(i item) string {
+	ts := i.typ.String()
+	if strings.Contains(ts, "<") {
```
Comment (juliusv) [inline promql/parse.go]: Wouldn't this also match `itemLSS`? The implicit string encoding to signal tokens without default representation is making me a bit nervous...

File: promql/parse.go [hunk 5]
```diff
@@ -354,6 +368,9 @@ func (p *parser) expr() Expr {
 
 		// Parse the next operand.
 		rhs := p.unaryExpr()
+		if rhs == nil {
+			p.errorf("could not find right-hand side of binary expression")
```
Comment (juliusv) [inline promql/parse.go]: keep it simple: "could not find" -> "missing" (and then s/of/in/ sounds better)

"missing right-hand side in binary expression"

File: promql/parse_test.go [hunk 6]
```diff
@@ -122,39 +124,77 @@ var testExpr = []struct {
...
 	}, {
-		input: "# just a comment\n\n", fail: true,
+		input:  "1+",
+		fail:   true,
+		errMsg: "could not find right-hand side of binary expression",
 	}, {
-		input: "1+", fail: true,
+		input:  ".",
+		fail:   true,
+		errMsg: "unrecognized character in statement: '.'",
```
Comment (juliusv) [inline promql/parse_test.go]: More like s/unrecognized/unexpected/?

s/statement/expression/?

File: promql/parse_test.go [hunk 7]
```diff
@@ -397,25 +437,45 @@ var testExpr = []struct {
...
 	}, {
-		input: "foo == on(bar) 10", fail: true,
+		input:  "foo == on(bar) 10",
+		fail:   true,
+		errMsg: "vector matching only allowed between vectors",
 	}, {
-		input: "foo and on(bar) group_left(baz) bar", fail: true,
+		input:  "foo and on(bar) group_left(baz) bar",
+		fail:   true,
+		errMsg: "no grouping allowed for AND and OR operations",
```
Comment (juliusv) [inline promql/parse_test.go]: We actually want to allow that... but better separate PR.

File: promql/parse_test.go [hunk 8]
```diff
@@ -470,31 +530,59 @@ var testExpr = []struct {
...
-		input: `{`, fail: true,
-	}, {
-		input: `}`, fail: true,
-	}, {
-		input: `some{`, fail: true,
-	}, {
-		input: `some}`, fail: true,
+		input:  `{`,
+		fail:   true,
+		errMsg: "unexpected EOF inside braces",
```
Comment (juliusv) [inline promql/parse_test.go]: EOF -> end of [expression|input]

(to make it sound less technical?)

File: promql/parse_test.go [hunk 9]
```diff
@@ -470,31 +530,59 @@ var testExpr = []struct {
...
 	}, {
-		input: `some_metric{a=b}`, fail: true,
+		input:  `}`,
+		fail:   true,
+		errMsg: "unrecognized character in statement: '}'",
```
Comment (juliusv) [inline promql/parse_test.go]: unrecognized -> unexpected

**Comment by juliusv** [comment]
Looks good otherwise.

File: promql/parse.go (hunk 3 above)
Comment (juliusv) [inline promql/parse.go]: Sounds good too.

File: promql/parse_test.go (hunk 6 above)
Comment (fabxc) [inline promql/parse_test.go]: The lexer has no notion of expression vs statement. It's a question of how much complexity/state we want to introduce for more specific error messages.

At some points it even makes sense to let the lex stuff we know to be wrong to have more state for errors in the parser - it's a fine line.

**Comment by fabxc** [comment]
All resolved.

File: promql/parse_test.go (hunk 6 above)
Comment (juliusv) [inline promql/parse_test.go]: Ok, then maybe just "character" instead of "character in statement", if it can happen both in a statement and an expression?

**Pull request merged**